static/js/components/email/hoc.js

```diff
@@ -0,0 +1,116 @@
+import { showDialog } from '../../actions/ui';
+import { emailValidation } from '../../lib/validation/profile';
+import { EMAIL_COMPOSITION_DIALOG } from './constants';
+import type {
+  EmailState,
+  EmailConfig
+} from '../../flow/emailTypes';
+
+export const withEmailDialog = R.curry(
+  (emailConfigs: {[key: string]: EmailConfig}, WrappedComponent) => {
+    class WithEmailDialog extends React.Component {
+      constructor(props) {
+        super(props);
+      }
+
+      getActiveEmailType(): ?string {
+        const { email } = this.props;
+        return !isNilOrBlank(email.currentlyActive) ? email.currentlyActive : undefined;
+      }
+
+      getActiveEmailState(): EmailState {
+        const { email } = this.props;
+        return email[email.currentlyActive];
+      }
+
+      openEmailComposer = (emailOpenParams: Object): void => {
+        const { dispatch } = this.props;
+        dispatch(startEmailEdit(emailOpenParams));
+        dispatch(showDialog(EMAIL_COMPOSITION_DIALOG));
+      };
+
+      updateEmailFieldEdit = R.curry((fieldName, e): void => {
+        const { dispatch, email: { currentlyActive } } = this.props;
+        let activeEmail = this.getActiveEmailState();
+        let inputsClone = R.clone(activeEmail.inputs);
+        inputsClone[fieldName] = e.target.value;
+        dispatch(updateEmailEdit({type: currentlyActive, inputs: inputsClone}));
+        if (!R.isEmpty(activeEmail.validationErrors)) {
+          let cloneErrors = emailValidation(inputsClone);
+          dispatch(updateEmailValidation({type: currentlyActive, errors: cloneErrors}));
+        }
+      });
+
+      closeAndClearEmailComposer = (): void => {
+        const { dispatch, email: { currentlyActive } } = this.props;
+        dispatch(clearEmailEdit(currentlyActive));
+        dispatch(hideDialog(EMAIL_COMPOSITION_DIALOG));
+      };
+
+      closeEmailComposerAndSend = (): void => {
+        const { dispatch, email: { currentlyActive } } = this.props;
+        let activeEmail = this.getActiveEmailState();
+        let errors = emailValidation(activeEmail.inputs);
+        dispatch(updateEmailValidation({type: currentlyActive, errors: errors}));
+        if (R.isEmpty(errors)) {
+          dispatch(
+            emailConfigs[currentlyActive].emailSendAction(activeEmail)
+          ).then(() => {
+            this.closeAndClearEmailComposer();
+          });
+        }
+      };
+
+      renderCompositionDialog(): ?React$Element<*> {
+        let { ui: { dialogVisibility } } = this.props;
+
+        let activeEmailType = this.getActiveEmailType();
+
+        if (!activeEmailType) return null;
+
+        return (
+          <EmailCompositionDialog
+            updateEmailFieldEdit={this.updateEmailFieldEdit}
+            closeAndClearEmailComposer={this.closeAndClearEmailComposer}
+            closeEmailComposerAndSend={this.closeEmailComposerAndSend}
+            dialogVisibility={dialogVisibility[EMAIL_COMPOSITION_DIALOG]}
```

Comment:
this PR makes sure that only one email dialog exists. the contents of that email dialog change based on the email state and `currentlyActive`